### ISSUE TITLE
FIXED IN #165 - Consider allowing custom encoders for JSON without monkey patching

### ISSUE THREAD

**Issue by posita** [fiorix/cyclone#133]
[MySQL-python](http://mysql-python.sourceforge.net/) (e.g., used from [twisted.enterprise.dbapi](https://twistedmatrix.com/documents/current/core/howto/rdbms.html)) uses [Decimal objects](https://docs.python.org/2/library/decimal.html#decimal-objects) to represent certain column values. These objects are not directly serializable in JSON.

There are various mechanisms that afford conversion to other objects (e.g., floats) before serialization, but there is a potential for loss of precision (see, e.g., [this](https://stackoverflow.com/questions/1960516/python-json-serialize-a-decimal-object), [this](https://stackoverflow.com/questions/7483363/python-mysqldb-returns-datetime-time-and-decimal), and [this](https://stackoverflow.com/questions/12898516/python-mysqldb-converters-isnt-working)). Some implementations support native conversion of these objects, but conversion without native support requires a custom encoder (`use_decimal=True` is not available like with Python 2.x's built-in `json`).

Currently, there is no easy way to use a custom encoder with cyclone's JSON-RPC library other than something like this:

``` python
#=========================================================================
def _monkeypatchcyclonejson():
    """
    Monkey patches cyclone.escape with a custom JSON encoder to handle
    Decimal objects. Adapted from: <http://tinyurl.com/kn3gpde>.
    """
    try:
        json = escape.json
    except AttributeError:
        try:
            json = escape.simplejson
        except AttributeError:
            json = None

    if json is not None:
        try:
            json.dumps([ Decimal('1.1') ], use_decimal=True)
        except TypeError:
            class fakefloat(float):
                def __init__(self, a_value):
                    float.__init__(self, a_value)
                    # Save the Decimal object
                    self._value = a_value

                def __repr__(self):
                    # Use the saved Decimal object for spitting out the
                    # repr value
                    return str(self._value)

            def decimal_default(a_obj):
                if isinstance(a_obj, Decimal):
                    return fakefloat(a_obj)

                raise TypeError(repr(a_obj) + ' is not JSON serializable')

            escape._json_encode = lambda v: json.dumps(v, default=decimal_default) # pylint: disable=protected-access
        else:
            escape._json_encode = lambda v: json.dumps(v, use_decimal=True) # pylint: disable=protected-access
```

There are probably more efficient way to code this. My only point is that it's still a monkey patch.

**Comment by dpnova** [comment]
We're adding some datetime stuff to our json request handlers. I actually feel the best way to handle this is to allow specifying custom json_encode and json_decode functions at the class level on the jsonrpc request handler. If you're at the point where you would like to change the behaviour that's a small amount of extra work to expect.

I'll create a PR for this shortly.

**Comment by fiorix** [comment]
I like this, and agree that it's in fact necessary.

**Comment by vltr** [comment]
+1. The serialization of some types (datetime, decimals, etc) are sometime a PITA. Perhaps a factory method, with the default json / simplejson impl as default if none registered?

**Comment by posita** [comment]
See @dpnova's proposed fix in #165.
